[docs] Adds Workflows documentation

jonsamp opened the pull request:
# Why

We are about to release a new product called Workflows. This PR details how to use Workflows with documentation.

# Test Plan

Read through the docs and make sure they make sense. You can read them in the following places:

- [/workflows/get-started](/workflows/get-started/)
- [/workflows/triggers](/workflows/triggers/)
- [/workflows/jobs](/workflows/jobs/)
- [/workflows/control-flow](/workflows/control-flow/)
- [/workflows/variables](/workflows/variables/)

jonsamp committed:
adds workflows documentation

jonsamp committed:
remove smart quotes

sjchmiela commented on docs/pages/workflows/variables.mdx:
```diff
@@ -0,0 +1,47 @@
+      fingerprint_hash_for_workflow: ${{ steps.fingerprint_step_id.outputs.fingerprint_hash }}
+      # @end #
+    steps:
+      - name: Set fingerprint hash
+			id: fingerprint_step_id
+			run: |
+				FINGERPRINT_HASH=$(npx expo-updates fingerprint:generate --platform ios | jq '.hash')
+                # @info Sets the `fingerprint_hash` as a variable #
+				set-output fingerprint_hash $FINGERPRINT_HASH
+                # @end #
+                # @info Exposes the `fingerprint_hash` variable to other steps in this job #
+			outputs: [fingerprint_hash]
+                # @end #
+      - name: Do something with the fingerprint hash
+			run: |
+                # @info Accesses the `fingerprint_hash` variable #
+				echo ${ steps.fingerprint_step_id.outputs.fingerprint_hash }
+                # @end #
```
```suggestion
      - name: Set fingerprint hash
        id: fingerprint_step_id
        # @info Exposes the `fingerprint_hash` variable to other steps in this job #
        outputs: [fingerprint_hash]
        # @end #
        run: |
          FINGERPRINT_HASH=$(npx expo-updates fingerprint:generate --platform ios | jq '.hash')
          # @info Sets the `fingerprint_hash` as a variable #
          set-output fingerprint_hash $FINGERPRINT_HASH
          # @end #
      - name: Do something with the fingerprint hash
          run: |
            # @info Accesses the `fingerprint_hash` variable #
            echo ${{ steps.fingerprint_step_id.outputs.fingerprint_hash }}
            # @end #
```

sjchmiela committed:
Update docs/pages/workflows/variables.mdx

sjchmiela commented on docs/pages/workflows/variables.mdx:
```diff
@@ -0,0 +1,47 @@
+          set-output fingerprint_hash $FINGERPRINT_HASH
+          # @end #
+      - name: Do something with the fingerprint hash
+          run: |
+            # @info Accesses the `fingerprint_hash` variable #
+            echo ${{ steps.fingerprint_step_id.outputs.fingerprint_hash }}
+            # @end #
```
```suggestion
        run: |
          # @info Accesses the `fingerprint_hash` variable #
          echo ${{ steps.fingerprint_step_id.outputs.fingerprint_hash }}
          # @end #
```

sjchmiela committed:
Update docs/pages/workflows/variables.mdx

amandeepmittal commented on docs/pages/workflows/get-started.mdx:
```diff
@@ -0,0 +1,143 @@
+
+<Collapsible summary="How do workflows compare to other CI services?">
+
+Workflows on EAS are designed to help you and your team release your app. It comes preconfigured with pre-packaged job types that help you build, submit, update, run Maestro tests, and more. It all runs on EAS, so you'll only have to manage one set of YAML files, and all of the artifacts from your job runs will appear on expo.dev.
```
```suggestion
Workflows on EAS are designed to help you and your team release your app. It comes preconfigured with pre-packaged job types that can build, submit, update, run Maestro tests, and more. All job types run on EAS, so you'll only have to manage one set of YAML files, and all the artifacts from your job runs will appear on [expo.dev](https://expo.dev/).
```

Let's make it a bit concise. It'll be an improvement to define pre-packaged jobs in the intro maybe.

amandeepmittal commented on docs/pages/workflows/get-started.mdx:
```diff
@@ -0,0 +1,143 @@
+
+Each workflow consists of three top-level elements:
+
+- `name`: defines the name of the workflow. For example: "Hello World!"
```
```suggestion
- `name`: defines the name of the workflow. For example: "Hello World"
```

Just to align with the example below, let's remove the exclamation mark from here.

amandeepmittal commented on docs/pages/workflows/get-started.mdx:
```diff
@@ -0,0 +1,143 @@
+- `on`: defines when this workflow should be triggered. For example: when pushing a new commit to a GitHub branch.
+- `jobs`: a sequence of jobs which can depend on and pass data between each other. For example: a job that runs a unit test followed by a job that builds your project into an app.
+
+Here's an example of a workflow that prints "Hello world!":
```
```suggestion
Here's an example of a workflow that prints "Hello, world":
```

Let's match the printed statement as described below.

amandeepmittal commented on docs/pages/workflows/get-started.mdx:
```diff
@@ -0,0 +1,143 @@
+
+- `name`: defines the name of the workflow. For example: "Hello World!"
+- `on`: defines when this workflow should be triggered. For example: when pushing a new commit to a GitHub branch.
+- `jobs`: a sequence of jobs which can depend on and pass data between each other. For example: a job that runs a unit test followed by a job that builds your project into an app.
```
```suggestion
- `jobs`: a sequence of jobs which can be interdependent and pass data between each other. For example, a job that runs a unit test followed by a job that builds your project into an app.
```

amandeepmittal commented on docs/pages/workflows/get-started.mdx:
```diff
@@ -0,0 +1,143 @@
+Each workflow consists of three top-level elements:
+
+- `name`: defines the name of the workflow. For example: "Hello World!"
+- `on`: defines when this workflow should be triggered. For example: when pushing a new commit to a GitHub branch.
```
```suggestion
- `on`: defines when this workflow should be triggered. For example, when pushing a new commit to a GitHub branch.
```

amandeepmittal commented on docs/pages/workflows/get-started.mdx:
```diff
@@ -0,0 +1,143 @@
+
+## Configure your project
+
+To enable workflows to run on events from GitHub, you'll need to install Expo's GitHub app and connect it to your project. Here's how:
```
```suggestion
To enable workflows to run on events from GitHub, you'll need to install Expo's GitHub app and connect it to your project:
```

The next two list items are self explanatory so we can omit this.

amandeepmittal commented on docs/pages/workflows/get-started.mdx:
```diff
@@ -0,0 +1,143 @@
+
+Once you do, you can see your workflow running on your project's [workflows page](https://expo.dev/accounts/[account]/projects/[projectName]/workflows).
+
+## Next steps
```
```suggestion
## Building blocks
```

Let's call this section building blocks instead of next steps since they (the links below) can be viewed in any order.

amandeepmittal commented on docs/pages/workflows/triggers.mdx:
```diff
@@ -0,0 +1,56 @@
+---
+title: Triggers
+description: Learn which triggers work with Workflows.
```
```suggestion
description: Learn which triggers are available to work with Workflows.
```

Let's improve this.

amandeepmittal commented on docs/pages/workflows/jobs.mdx:
```diff
@@ -0,0 +1,84 @@
+
+Workflows sequence jobs together to automate your development and release processes. You can run two types of jobs: pre-packaged jobs and custom jobs.
+
+Pre-packaged jobs are provided by Expo and meet common use cases for developers, like building, submitting, and running end to end test. Since they are provided by Expo, they are continuously maintained and updated.
```
```suggestion
Expo provides pre-packaged jobs that meet common use cases for developers, such as building, submitting, and running end-to-end tests. Since they are provided by Expo, they are continuously maintained and updated.
```

amandeepmittal commented on docs/pages/workflows/jobs.mdx:
```diff
@@ -0,0 +1,84 @@
+
+Pre-packaged jobs are provided by Expo and meet common use cases for developers, like building, submitting, and running end to end test. Since they are provided by Expo, they are continuously maintained and updated.
+
+Custom jobs are jobs that you write yourself. They are useful for any use case that isn't met by the pre-packaged jobs.
```
```suggestion
Custom jobs are jobs that you write yourself. They are useful for any use case not met by the pre-packaged jobs.
```

amandeepmittal commented on docs/pages/workflows/jobs.mdx:
````diff
@@ -0,0 +1,84 @@
+    type: update
+```
+
+## End-to-end test with Maestro
````
```suggestion
## End-to-end tests with Maestro
```

amandeepmittal commented on docs/pages/workflows/control-flow.mdx:
```diff
@@ -0,0 +1,69 @@
+
+## `needs`
+
+You can add an list of job names to the `needs` keyword on a job. If a job in the list fails, the job will be skipped.
```
```suggestion
You can add a list of previous job names using the `needs` keyword on a job. If any job from that list fails, the current job will be skipped.
```

At first it was a bit confusing to read this sentence to quickly understand what was happening. In the last sentence, I am not sure if the current job is skipped or the job failed from the list is skipped.

Assuming that the only previous job names can be added to the job using the `needs` keyword, let's explicitly mention that to improve clarity.

Please correct me if my assumption is wrong and feel free to update the sentence accordingly :)

amandeepmittal commented on docs/pages/workflows/control-flow.mdx:
```diff
@@ -0,0 +1,69 @@
+
+## `after`
+
+You can add a list of job names to the `after` keyword on a job. The jobs in the list will run after the job completes no matter if it succeeded or failed.
```
```suggestion
You can add a list of previous job names using the `after` keyword on a job. Any job in the list will run after the current job completes, no matter if it succeeded or failed.
```

amandeepmittal commented on docs/pages/workflows/variables.mdx:
```diff
@@ -0,0 +1,47 @@
+description: Learn how to use variables in Workflows.
+---
+
+You can use variables in your workflows to pass data between jobs or steps. Variables are useful for conditionals on subsequent jobs or afftecting the output of a workflow.
```
```suggestion
Variables can be used in workflows to pass data between jobs or steps. They are useful for conditionals on subsequent jobs or affecting the output of a workflow.
```

amandeepmittal commented on docs/pages/workflows/variables.mdx:
````diff
@@ -0,0 +1,47 @@
+
+You can set variables in the workflow file with the `set-output` function:
+
+```yaml .eas/workflows/variables.yaml
+jobs:
+  fingerprint:
+    outputs:
+      # @info Sets the `fingerprint_hash_for_workflow` as a variable #
+      fingerprint_hash_for_workflow: ${{ steps.fingerprint_step_id.outputs.fingerprint_hash }}
+      # @end #
+    steps:
+      - name: Set fingerprint hash
+        id: fingerprint_step_id
+        # @info Exposes the `fingerprint_hash` variable to other steps in this job #
+        outputs: [fingerprint_hash]
+        # @end #
+        run: |
+          FINGERPRINT_HASH=$(npx expo-updates fingerprint:generate --platform ios | jq '.hash')
+          # @info Sets the `fingerprint_hash` as a variable #
+          set-output fingerprint_hash $FINGERPRINT_HASH
+          # @end #
+      - name: Do something with the fingerprint hash
+        run: |
+          # @info Accesses the `fingerprint_hash` variable #
+          echo ${{ steps.fingerprint_step_id.outputs.fingerprint_hash }}
+          # @end #
+    maybe_build:
+      needs: [fingerprint]
+      # @info Accesses the `fingerprint_hash_for_workflow` variable #
+      if: ${{ needs.fingerprint.outputs.fingerprint_hash_for_workflow != '1234' }}
+      # @end #
+      type: build
+      params:
+        platform: ios
+```
````
````suggestion
```yaml .eas/workflows/variables.yaml
jobs:
  fingerprint:
    outputs:
      # @info Sets the <CODE>fingerprint_hash_for_workflow</CODE> as a variable #
      fingerprint_hash_for_workflow: ${{ steps.fingerprint_step_id.outputs.fingerprint_hash }}
      # @end #
    steps:
      - name: Set fingerprint hash
        id: fingerprint_step_id
        # @info Exposes the <CODE>fingerprint_hash</CODE> variable to other steps in this job #
        outputs: [fingerprint_hash]
        # @end #
        # @info Sets the <CODE>fingerprint_hash</CODE> as a variable #
        run: |
          FINGERPRINT_HASH=$(npx expo-updates fingerprint:generate --platform ios | jq '.hash')
          set-output fingerprint_hash $FINGERPRINT_HASH
        # @end #
      - name: Do something with the fingerprint hash
        # @info Accesses the <CODE>fingerprint_hash</CODE variable #
        run: |
          echo ${{ steps.fingerprint_step_id.outputs.fingerprint_hash }}
        # @end #
    maybe_build:
      needs: [fingerprint]
      # @info Accesses the <CODE>fingerprint_hash_for_workflow</CODE> variable #
      if: ${{ needs.fingerprint.outputs.fingerprint_hash_for_workflow != '1234' }}
      # @end #
      type: build
      params:
        platform: ios
```
````

The formatting of code comments for sequential jobs under run was broken:

![CleanShot 2024-11-16 at 22 26 25@2x](https://github.com/user-attachments/assets/176fbf28-8091-4df4-b0ca-80024efa89b9)


For now, I could only fix them if the comment surrounds the whole line (including the step keyword)

![CleanShot 2024-11-16 at 22 25 19@2x](https://github.com/user-attachments/assets/86171566-dd48-4648-980f-d809979ef28a)

Also, let's use `<CODE>` to highlight code syntax instead of back ticks since back ticks do not work in annotated comments to have the same affect

![CleanShot 2024-11-16 at 22 23 43@2x](https://github.com/user-attachments/assets/b29e4b4b-bbe3-4226-af10-b9dd6e02cf30)

amandeepmittal commented on docs/pages/workflows/variables.mdx:
```diff
@@ -0,0 +1,47 @@
+---
+title: Variables
+description: Learn how to use variables in Workflows.
```
```suggestion
description: Learn how to pass data between jobs or steps or set conditionals using variables in Workflows.
```

Let's make it more descriptive.

jonsamp committed:
Apply suggestions from code review

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

jonsamp committed:
[docs] Adds workflows to /guides (#32999)

# Why

This adds links to workflows docs under /guides, in the EAS and above
EAS Build. I chose this spot since workflows create builds, submissions,
updates, etc.

# Screenshot

<img width="328" alt="Screenshot 2024-11-17 at 11 56 11 AM"
src="https://github.com/user-attachments/assets/eae8f4ad-c896-4381-ab35-e1954a060601">


# Test Plan

Make sure that the workflows docs show up under the /guides section in
the docs.
